stdlib/public/runtime/ReflectionMirror.mm

```diff
@@ -0,0 +1,727 @@
+  // contained value.
+  // TODO: Should look through existential metatypes too, but it doesn't
+  // really matter yet since we don't have any special mirror behavior for
+  // concrete metatypes yet.
+  while (T->getKind() == MetadataKind::Existential) {
+    auto existential
+      = static_cast<const ExistentialTypeMetadata *>(T);
+
+    // Unwrap the existential container.
+    T = existential->getDynamicType(Value);
+    Value = existential->projectValue(Value);
+
+    // Existential containers can end up nested in some cases due to generic
+    // abstraction barriers.  Repeat in case we have a nested existential.
+  }
+  return std::make_tuple(T, Value);
+}
+
+static bool loadSpecialReferenceStorage(OpaqueValue *fieldData,
+                                        const FieldType fieldType,
+                                        Any *outValue) {
+  // isWeak() implies a reference type via Sema.
+  if (!fieldType.isWeak())
+    return false;
+
+  auto type = fieldType.getType();
+  assert(type->getKind() == MetadataKind::Optional);
+
+  auto weakField = reinterpret_cast<WeakReference *>(fieldData);
+  auto strongValue = swift_unknownWeakLoadStrong(weakField);
+
+  // Now that we have a strong reference, we need to create a temporary buffer
+  // from which to copy the whole value, which might be a native class-bound
+  // existential, which means we also need to copy n witness tables, for
+  // however many protocols are in the protocol composition. For example, if we
+  // are copying a:
+  // weak var myWeakProperty : (Protocol1 & Protocol2)?
+  // then we need to copy three values:
+  // - the instance
+  // - the witness table for Protocol1
+  // - the witness table for Protocol2
+
+  auto weakContainer =
+    reinterpret_cast<WeakClassExistentialContainer *>(fieldData);
+
+  // Create a temporary existential where we can put the strong reference.
+  // The allocateBuffer value witness requires a ValueBuffer to own the
+  // allocated storage.
+  ValueBuffer temporaryBuffer;
+
+  auto temporaryValue = reinterpret_cast<ClassExistentialContainer *>(
+      type->allocateBufferIn(&temporaryBuffer));
+
+  // Now copy the entire value out of the parent, which will include the
+  // witness tables.
+  temporaryValue->Value = strongValue;
+  auto valueWitnessesSize = type->getValueWitnesses()->getSize() -
+                            sizeof(WeakClassExistentialContainer);
+  memcpy(temporaryValue->getWitnessTables(), weakContainer->getWitnessTables(),
+         valueWitnessesSize);
+
+  outValue->Type = type;
+  auto *opaqueValueAddr = type->allocateBoxForExistentialIn(&outValue->Buffer);
+  type->vw_initializeWithCopy(opaqueValueAddr,
+                              reinterpret_cast<OpaqueValue *>(temporaryValue));
+
+  type->deallocateBufferIn(&temporaryBuffer);
+  
+  return true;
+}
+
+// Get a field name from a doubly-null-terminated list.
+static const char *getFieldName(const char *fieldNames, size_t i) {
```

Comment:
Generally in LLVM style we just use unsigned instead of size_t. But if people are using size_t in the runtime, better to be consistent. I would see what people are using.